File: apps/api/app/api/v1/codex_progressions.py
import uuid
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db
from app.schemas.codex_progression import (
    CodexProgressionCreate,
    CodexProgressionRead,
    CodexProgressionUpdate,
)
from app.services.crud_codex_progression import (
    create_codex_progression,
    delete_codex_progression,
    get_codex_progression,
    list_codex_progressions,
    update_codex_progression,
)

router = APIRouter(prefix="/codex-progressions", tags=["codex_progressions"])


@router.post("", response_model=CodexProgressionRead, status_code=status.HTTP_201_CREATED)
async def create(
    data: CodexProgressionCreate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexProgressionRead:
    return await create_codex_progression(db, data)


@router.get("", response_model=list[CodexProgressionRead])
async def list_all(
    entity_type: str = Query(...),
    entity_id: uuid.UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[CodexProgressionRead]:
    return await list_codex_progressions(db, entity_type, entity_id)


@router.get("/{progression_id}", response_model=CodexProgressionRead)
async def get_one(
    progression_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexProgressionRead:
    progression = await get_codex_progression(db, progression_id)
    if progression is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex progression not found"
        )
    return progression


@router.patch("/{progression_id}", response_model=CodexProgressionRead)
async def update(
    progression_id: uuid.UUID,
    data: CodexProgressionUpdate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexProgressionRead:
    progression = await get_codex_progression(db, progression_id)
    if progression is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex progression not found"
        )
    return await update_codex_progression(db, progression, data)


@router.delete("/{progression_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    progression_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> None:
    progression = await get_codex_progression(db, progression_id)
    if progression is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Codex progression not found"
        )
    await delete_codex_progression(db, progression)

File: apps/web/lib/api/codex-progressions.ts
/**
 * Typed endpoint functions for the CodexProgression resource (Progresszió tab).
 * FLAT router (`apps/api/app/api/v1/codex_progressions.py`, prefix
 * `/codex-progressions` under `/api/v1`):
 *   GET    /codex-progressions?entity_type=&entity_id=  → CodexProgressionRead[]
 *   POST   /codex-progressions                          → CodexProgressionRead (201)
 *   GET    /codex-progressions/{id}                     → CodexProgressionRead
 *   PATCH  /codex-progressions/{id}                     → CodexProgressionRead
 *   DELETE /codex-progressions/{id}                     → 204
 *
 * A progression records an entity's state change, optionally anchored to a
 * chapter and/or scene (anchorless = project-global baseline). The backend CRUD
 * lists by `created_at desc`; STORY ordering (the order the AI-context
 * linearization uses) is computed client-side from the book tree — see
 * `components/codex/progression-tab.tsx`. Each function validates its response
 * with Zod — backend drift throws rather than silently flowing a wrong shape.
 */
import { apiFetch } from "./client";
import {
  codexProgressionCreateSchema,
  codexProgressionListSchema,
  codexProgressionReadSchema,
  codexProgressionUpdateSchema,
  type CodexProgressionCreate,
  type CodexProgressionRead,
  type CodexProgressionUpdate,
} from "./types";

/** List the progressions recorded for one entity (`entity_type` + id). */
export async function listCodexProgressions(
  entityType: string,
  entityId: string,
): Promise<CodexProgressionRead[]> {
  const query = new URLSearchParams({
    entity_type: entityType,
    entity_id: entityId,
  });
  const data = await apiFetch<unknown>(`/codex-progressions?${query}`);
  return codexProgressionListSchema.parse(data);
}

/** Create a progression. The payload is validated before it is sent. */
export async function createCodexProgression(
  input: CodexProgressionCreate,
): Promise<CodexProgressionRead> {
  const body = codexProgressionCreateSchema.parse(input);
  const data = await apiFetch<unknown>(`/codex-progressions`, {
    method: "POST",
    body,
  });
  return codexProgressionReadSchema.parse(data);
}

/** Patch a progression's anchor and/or note. */
export async function updateCodexProgression(
  progressionId: string,
  patch: CodexProgressionUpdate,
): Promise<CodexProgressionRead> {
  const body = codexProgressionUpdateSchema.parse(patch);
  const data = await apiFetch<unknown>(
    `/codex-progressions/${progressionId}`,
    { method: "PATCH", body },
  );
  return codexProgressionReadSchema.parse(data);
}

/** Delete a progression (the backend answers 204; apiFetch returns null). */
export async function deleteCodexProgression(
  progressionId: string,
): Promise<void> {
  await apiFetch<unknown>(`/codex-progressions/${progressionId}`, {
    method: "DELETE",
  });
}
